iree/compiler/Dialect/Flow/Conversion/TensorToFlow/ConvertTensorToFlow.cpp

```diff
@@ -0,0 +1,383 @@
+
+    if (staticStride != 1) return false;
+    // The offsets and sizes dont have to be static for all dimensions. When
+    // `fullSlices` is true, the offset and sizes can be dynamic. But many
+    // cases, the dynamic offset/size value is obtained by computing from
+    // another tensor which lives on the device. To avoid host-round tripping
+    // enforce that offset/size is also static.
+    if (staticSize == ShapedType::kDynamicSize) return false;
+    if (staticOffset == ShapedType::kDynamicStrideOrOffset) return false;
+
+    if (fullSlices == false) {
+      if (staticSize != 1) return false;
+    } else {
+      if (!(staticOffset == 0 && staticSize != ShapedType::kDynamicSize &&
+            baseShape[dim - 1] != ShapedType::kDynamicSize &&
+            staticSize == baseShape[dim - 1]))
+        fullSlices = false;
+    }
+  }
+  return true;
+}
+
+/// Returns the `Value`s for a list of `OpFoldResult` by generating std.constant
+/// ops for the static values.
+static SmallVector<Value, 4> getAsValues(
+    OpBuilder &b, Location loc, ArrayRef<OpFoldResult> valueOrAttrList) {
+  SmallVector<Value, 4> values;
+  for (auto valueOrAttr : valueOrAttrList) {
+    if (auto attr = valueOrAttr.dyn_cast<Attribute>()) {
+      values.push_back(
+          b.create<ConstantIndexOp>(loc, attr.cast<IntegerAttr>().getInt()));
+    } else {
+      values.push_back(valueOrAttr.get<Value>());
+    }
+  }
+  return values;
+}
+
+/// Gets the list of non-static values from a list of `OpFoldResult`.
+static SmallVector<Value, 4> getDynamicValues(
+    ArrayRef<OpFoldResult> valueOrAttrList) {
+  SmallVector<Value, 4> dynamicDims;
+  for (auto valueOrAttr : valueOrAttrList) {
+    if (auto value = valueOrAttr.dyn_cast<Value>()) {
+      dynamicDims.push_back(value);
+    }
+  }
+  return dynamicDims;
+}
+
+/// Get shape of the tensor given the sizes as a list of `OpFoldResult`.
+static SmallVector<int64_t, 4> getShapeFromSizes(
+    ArrayRef<OpFoldResult> valueOrAttrList) {
+  return llvm::to_vector<4>(llvm::map_range(
+      valueOrAttrList, [&](OpFoldResult valueOrAttr) -> int64_t {
+        if (auto attr = valueOrAttr.dyn_cast<Attribute>()) {
+          return attr.cast<IntegerAttr>().getInt();
+        }
+        return ShapedType::kDynamicSize;
+      }));
+}
+
+/// Generates `memref.dim` operations to get the dynamic sizes of a value `v`.
+static SmallVector<Value, 4> getDynamicDimValues(OpBuilder &b, Location loc,
+                                                 Value v) {
+  SmallVector<Value, 4> dynamicDims;
+  for (auto dim : llvm::enumerate(v.getType().cast<ShapedType>().getShape())) {
+    if (dim.value() != ShapedType::kDynamicSize) continue;
+    dynamicDims.push_back(b.createOrFold<tensor::DimOp>(loc, v, dim.index()));
+  }
+  return dynamicDims;
+}
+
+/// Convert tensor.insert_slice ops into flow.tensor.update ops where possible.
+struct ConvertTensorInsertSlicePattern
```

Comment:
Why are these changed to a conversion pattern rewrite? AFAIK, you can still keep this as a `OpRewritePattern` and use it in the dialect conversion. In general better to avoid dialect conversion unless you absolutely need it (and you only absolutely need it if you are changing types, which is not the case here since all types are `tensor`s)

Reply:
Ah, I'll try changing them all back to `OpRewritePatterns`. I wasn't really sure which parts were load bearing in the existing implementations.

Reply:
The `CastOp` conversion was using dialect conversion features and goes from static<->dynamic dimensions: 
https://github.com/google/iree/blob/3554868e9945cb48fe1ef2b2adfec96253c40171/iree/compiler/InputConversion/Common/ConvertUpstreamToIREE.cpp#L26-L41

Is conversion actually needed there or is this possible to express using an `OpRewritePattern`?

Reply:
ah, thats painful. Typically to switch between static <-> dynamic you would use a `tensor.cast` but that doesnt work for `tensor.cast` itself.

Maybe we dont move the pattern into a common place? So `tensor.cast` -> whatever for input conversion stays as is. What is your need for `tensor.cast` conversion?

Reply:
Hmm.. well, I pushed a commit that seems to work changing back to `OpRewritePattern` for all of these. I don't have a particular use for the `tensor.cast` conversion at this stage, but would prefer to group it with the other `tensor`->`flow` dialect patterns if possible.

Reply:
Ok, ill take a look. If keeping it as `OpRewritePattern` doesnt hurt, then its ok.